### PR TITLE
Handle resource cleanup on termination within the inventory control stream

### PR DESCRIPTION
**Pull request by rosstimothy**
Historically, each component responsible for managing a particular resource has been responsible for deleting said resources if Teleport was shutting down for good. This is done to provide a better user experience by trying to eliminate the chance of a stale resource from being present for the full TTL of the backend item. However, the mechanism to do so requires the process responsible for the resources to delete each resource individually. For dynamic resources this could require several thousand Delete RPCs during shutdown. Since the shutdown process is time bound this could result in some deletions from being successful and lead to the same stale resource issues. This also poses a problem and races with any inbound heartbeats being sent via the inventory control stream. All resource deletion mechanism are unary RPCs  that get processed outside of the inventory control stream, which without careful coordination could result in all the deletions occurring before any in flight heartbeats are processed by the inventory control stream.

In an attempt to simplify the deletion process a new `UpstreamInventoryGoodbye` message has been added to the inventory control stream in order to allow deletion to occur in-band. Instead of sending individual delete RPCs when a process is terminating it can now send a single `UpstreamInventoryGoodbye` via the inventory control stream. Any control streams that receive a GoodBye prior to being terminated indicate to auth that it should remove all resources that were being represented by said stream. Additionally, the `DownstreamInventoryHello` was updated to include supported capabilities to better coordinate backward compatibility and supported features by the auth server. This allows the agent to get a better understanding of what the auth server supports with out having to do a series of version checks.

### CODE REVIEW

File: lib/inventory/controller.go [hunk 1]
```diff
@@ -371,6 +372,11 @@ func (c *Controller) handleControlStream(handle *upstreamHandle) {
 				}
 			case proto.UpstreamInventoryPong:
 				c.handlePong(handle, m)
+			case proto.UpstreamInventoryDelete:
+				if err := c.handleTermination(handle, m); err != nil {
+					handle.CloseWithError(err)
+					return
+				}
```
Comment (fspmarshall) [inline lib/inventory/controller.go:379]: If this message is going to effectively terminate the stream, consider naming it `UpstreamInventoryGoodbye` and having the deletion behavior be controlled by a field.  That way we can easily add any other future "on exit" stuff to this message.

File: lib/inventory/inventory.go [hunk 2]
```diff
@@ -305,11 +310,19 @@ func (h *downstreamHandle) Sender() <-chan DownstreamSender {
...
+	return h.shutdownContext
 }
 
 func (h *downstreamHandle) Close() error {
-	h.cancel()
+	h.closeCancel()
+	return nil
+}
+
+func (h *downstreamHandle) Shutdown() error {
```
Comment (fspmarshall) [inline lib/inventory/inventory.go:321]: nit: `Shutdown` should take a context, and should force-close and exit if the context is canceled.

**Review by fspmarshall — APPROVED** [review]

File: lib/inventory/controller.go [hunk 3]
```diff
@@ -491,6 +503,13 @@ func (c *Controller) handlePingRequest(handle *upstreamHandle, req pingRequest)
 }
 
 func (c *Controller) handleHeartbeatMsg(handle *upstreamHandle, hb proto.InventoryHeartbeat) error {
+	// If a goodbye was received that indicated active resource
+	// clean up should occur, then ignore processing any heartbeats
+	// that may have been emitted after the goodbye.
+	if handle.goodbye != nil && handle.goodbye.DeleteResources {
+		return nil
+	}
```
Comment (fspmarshall) [inline lib/inventory/controller.go:511]: nit: I could go either way on this, but my gut feeling is that this condition is added brittleness/complexity without much added upside.  I'm partial to the idea of the goodbye not changing anything about the state-machine's behavior until the stream actually exits.  That feels like a simpler/more reliable system to me.  I do see the upside of avoiding unnecessary writes tho, so take it or leave it.

File: lib/inventory/inventory.go [hunk 4]
```diff
@@ -313,6 +300,17 @@ func (h *downstreamHandle) Close() error {
 	return nil
 }
 
+func (h *downstreamHandle) Goodbye(ctx context.Context) error {
```
Comment (fspmarshall) [inline lib/inventory/inventory.go:303]: nit: godoc.  More specifically, we probably want the godoc to make a clear explanation of the conceptual difference between sending the goodbye message and actually closing.

File: lib/srv/app/server.go [hunk 5]
```diff
@@ -465,6 +465,17 @@ func (s *Server) close(ctx context.Context) error {
...
+	select {
+	case sender := <-s.c.InventoryHandle.Sender():
+		// Manual deletion per app is only required if the auth server
+		// doesn't support actively cleaning up app resources when the
+		// inventory control stream is terminated during shutdown.
+		if capabilities := sender.Hello().Capabilities; capabilities != nil {
+			shouldDeleteApps = shouldDeleteApps && !capabilities.AppCleanup
+		}
+	default:
+	}
```
Comment (fspmarshall) [inline lib/srv/app/server.go:477]: non-blocking selects on `Sender()` channel are pretty racy.  It'll fail if either a concurrent goroutine is also pulling the sender, or if a downstream message arrives concurrently.  This should either be a blocking select with a timeout, or we should add some kind of official mechanism of doing non-blocking sender acquisition (e.g. `GetSender() (s DownstreamSender, ok bool)`.

**Review by fspmarshall — APPROVED** [review]

File: api/proto/teleport/legacy/client/proto/authservice.proto [hunk 6]
```diff
@@ -2367,6 +2414,14 @@ message InventoryHeartbeat {
   types.AppServerV3 AppServer = 2;
 }
 
+// UpstreamInventoryGoodbye informs the upstream service that instance
+// is terminating
+message UpstreamInventoryGoodbye {
+  // DeleteResources indicates that any heartbeats recevied from
```
Comment (espadolini) [inline api/proto/teleport/legacy/client/proto/authservice.proto:2420]:
```suggestion
  // DeleteResources indicates that any heartbeats received from
```

File: lib/inventory/controller.go [hunk 7]
```diff
@@ -319,6 +320,15 @@ func (c *Controller) handleControlStream(handle *upstreamHandle) {
 	}
 
 	defer func() {
+		if handle.goodbye != nil && handle.goodbye.DeleteResources {
```
Comment (espadolini) [inline lib/inventory/controller.go:323]:
```suggestion
		if handle.goodbye.GetDeleteResources() {
```

File: lib/inventory/controller.go [hunk 8]
```diff
@@ -319,6 +320,15 @@ func (c *Controller) handleControlStream(handle *upstreamHandle) {
...
+			log.WithField("apps", len(handle.appServers)).Debug("Cleaning up resources in response to instance termination")
+			for _, app := range handle.appServers {
+				if err := c.auth.DeleteApplicationServer(c.closeContext, apidefaults.Namespace, app.resource.GetHostID(), app.resource.GetName()); err != nil && !trace.IsNotFound(err) {
```
Comment (espadolini) [inline lib/inventory/controller.go:326]: Should we conditionally delete based on our last known heartbeat revision (if we have that anywhere)?

File: lib/inventory/controller.go [hunk 9]
```diff
@@ -371,6 +381,8 @@ func (c *Controller) handleControlStream(handle *upstreamHandle) {
 				}
 			case proto.UpstreamInventoryPong:
 				c.handlePong(handle, m)
+			case proto.UpstreamInventoryGoodbye:
+				handle.goodbye = &m
 			default:
 				log.Warnf("Unexpected upstream message type %T on control stream of server %q.", m, handle.Hello().ServerID)
 				handle.CloseWithError(trace.BadParameter("unexpected upstream message type %T", m))
```
Comment (espadolini) [inline lib/inventory/controller.go:388]: Should the client only send the goodbye message if the server has advertised support for any of the cleanups? Should we have a flag for "supports goodbye message"?

**Review by espadolini — APPROVED** [review]
What's the behavior of a grpc stream if we send a message and then immediately terminate the stream or connection? Do we have a chance for the message to be actually sent at all, or is it almost always forgotten in the connection buffers?

**Comment by tigrato** [comment]
> What's the behavior of a grpc stream if we send a message and then immediately terminate the stream or connection? Do we have a chance for the message to be actually sent at all, or is it almost always forgotten in the connection buffers?

It's likely sent but the server handles the context cancellation first. 
If you send a message + calling sending close from the client, you must call recv even if the server won't send anything else otherwise the connection is terminated and server connection context is cancelled

**Review by fspmarshall — APPROVED** [review]

**Pull request merged**